### PR TITLE
Fix #43 and tailable cursors

### LINKED ISSUE #43: Add database option support for CRUD functions
distinct/find/insert_one/insert_many/delete_one/delete_many/replace_one/update_one/update_many/save_one/save_many including MongoCursors should support :database option to specify a database to talk with.

### PR DESCRIPTION
**Pull request by chowevtfy**
There's apparently also a problem with DBConnection not actually calling `checkin` and `checkout` that is fixed in one of the commits with a simple workaround. Further investigation on that is necessary, but this does let people try tailable cursors and the database option.

### CODE REVIEW

File: lib/mongo/cursor.ex [hunk 1]
```diff
@@ -59,7 +59,11 @@ defmodule Mongo.Cursor do
 
           case Mongo.get_more(conn, coll, cursor, opts) do
             {:ok, %{cursor_id: cursor, docs: []}} ->
-              {:halt, state(state, cursor: cursor)}
+              if opts[:tailable_cursor] == true do
```
Comment (ericmj) [inline lib/mongo/cursor.ex:62]: Drop the `== true`? Or move `opts[:tailable_cursor]` into a variable and make it a guard on the case?

File: lib/mongo/protocol.ex [hunk 2]
```diff
@@ -112,7 +112,13 @@ defmodule Mongo.Protocol do
   end
 
   def handle_execute(%Mongo.Query{action: action, extra: extra}, params, opts, s) do
-    handle_execute(action, extra, params, opts, s)
+    new_s = %{s | database: Keyword.get(opts, :database, s.database)}
+    :ok = :inet.setopts(new_s.socket, [active: false])
+    with {:ok, reply, new_s} <-
+           handle_execute(action, extra, params, opts, new_s) do
```
Comment (ericmj) [inline lib/mongo/protocol.ex:118]: Don't split this into two lines.

**Comment by chowevtfy** [comment]
Having problems with streams in this PR. Since the stream never closes, it's not possible to stop it. If you kill the process where it is running, it will crash the Mongo connection.

**Comment by ericmj** [comment]
Can you create a failing test for what happens when you kill the process so I can take a look?

**Comment by ghost** [comment]
Yes, I can, sorry. I'll get to this once I'm able.

**Comment by ericmj** [comment]
@cjhowe Ping? Please rebase against master.

**Comment by chowevtfy** [comment]
Considering splitting this PR. I've been using this implementation of the database option and haven't had any problems with it, but the tailable cursors are going to need more work. I think the database option should be split into a new PR.

**Comment by ericmj** [comment]
What's the status of this PR compared to the replica set one? Is this something we can merge before?

**Comment by ghost** [comment]
No, this is gonna have to wait until after the replica set PR. One thing is that this issue is somewhat dependent on whether we end up pulling in GenStage as a dependency, which would make the solution trivial. I'm not sure this PR is viable at all anymore. I think tailable await provides a decent workaround for now.

**Pull request closed without merging**